elastic/elastic-package
# Support routing_rules.yml and ingest pipeline test for reroute processor

## What does this PR do?

This PR does two things: 
1. Support `routing_rules.yml` file when running pipeline test using command `elastic-package test pipeline -d <datastream> --generate -v`.

`routing_rules.yml` now can be read and take into account when running the test pipeline command. The file is read as yml and convert into reroute processors. Then these reroute processors are attached under the `processors` section from the ingest pipeline.

For example: 
```
- source_dataset: awsfirehose.log
  rules:
    - target_dataset: aws.cloudtrail
      if: ctx['aws.cloudwatch.log_stream'].contains('CloudTrail')
      namespace:
        - "{{labels.data_stream.namespace}}"
        - default
```

`routing_rules.yml` looks like the example above will be translated into:
```
- reroute:
      tag: awsfirehose.log
      if: ctx['aws.cloudwatch.log_stream'].contains('CloudTrail')
      dataset: aws.cloudtrail
      namespace: ["{{labels.data_stream.namespace}}", default]
``` 

2. Support ingest pipeline test for reroute processors when they are added under `elasticsearch/ingest_pipeline` directly.

- This PR added the index name for testing based on the dataset so we don't see error like this:

```
[0] unexpected pipeline error: invalid data stream name: [_index]; must follow naming scheme <type>-<dataset>-<namespace>
```

- When checking the value for `data_stream.dataset`, it's read from the reroute processors for the new target dataset. Because reroute processor will change the value of this field.

## How to test this PR locally
I've been testing this PR locally with https://github.com/elastic/integrations/pull/7146. After `make build` elastic package locally, I ran `elastic-package test pipeline -d log --generate -v` and under `integrations/packages/awsfirehose/data_stream/log/_dev/test/pipeline/test-cloudtrail.json-expected.json`, you can see the `data_stream.dataset` value changed to `aws.cloudtrail` and `data_stream.namespace` changed to `default`. These are the two fields I defined under `reroute` processor:
```
  - reroute:
      if: ctx['aws.cloudwatch.log_stream'].contains('CloudTrail')
      dataset: aws.cloudtrail
      namespace: default
```

Same thing works for defining it under `routing_rules.yml`
```
- source_dataset: awsfirehose.log
  rules:
    - target_dataset: aws.cloudtrail
      if: ctx['aws.cloudwatch.log_stream'].contains('CloudTrail')
      namespace:
        - "{{labels.data_stream.namespace}}"
        - default
```
